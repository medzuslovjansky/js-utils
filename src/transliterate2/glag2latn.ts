import { createReplacer } from './createReplacer';
import { Glagolitic } from '../constants';
import { replaceByWord } from './replaceByWord';

const punctuate = createReplacer({
  ';': '?',
  '։': '.',
  '·': ',',
  '⁖': '(',
  '჻': ')',
  '⸪': '"',
  '··': '!',
  '⸬': ':',
  '⁘': ';',
  '···': '...',
});

const softener = createReplacer({
  dь: 'ď',
  lь: 'ľ',
  nь: 'ń',
  rь: 'ŕ',
  sь: 'ś',
  tь: 'ť',
  zь: 'ź',
  ьr: 'ŕ',
  ȯr: 'r',
});

const yerify = (word: string) =>
  word.replace(/([dlnrstz])ь([cčklnr])/, '$1ė$2');

const deglagolify = createReplacer({
  [Glagolitic.Azu]: 'a',
  [Glagolitic.Otu]: 'å',
  [Glagolitic.Buky]: 'b',
  [Glagolitic.Tsi]: 'c',
  [Glagolitic.Chrivi]: 'č',
  [Glagolitic.Chrivi + Glagolitic.Yeri]: 'ć',
  [Glagolitic.Shta]: 'ć',
  [Glagolitic.Dobro]: 'd',
  [Glagolitic.Djervi]: 'đ',
  [Glagolitic.Dzelo]: 'ѕ',
  [Glagolitic.Yestu]: 'e',
  [Glagolitic.Yati]: 'ě',
  [Glagolitic.Small_Yus]: 'ę',
  [Glagolitic.Yeri]: 'ь',
  [Glagolitic.Fritu]: 'f',
  [Glagolitic.Glagoli]: 'g',
  [Glagolitic.Heru]: 'h',
  [Glagolitic.Spidery_Ha]: 'h',
  [Glagolitic.I]: 'i',
  [Glagolitic.Izhe]: 'j',
  [Glagolitic.Initial_Izhe]: 'j',
  [Glagolitic.Trokutasti_A]: 'ja',
  [Glagolitic.Iotated_Small_Yus]: 'ję',
  [Glagolitic.Yo]: 'jo',
  [Glagolitic.Yu]: 'ju',
  [Glagolitic.Iotated_Big_Yus]: 'jų',
  [Glagolitic.Kako]: 'k',
  [Glagolitic.Ljudije]: 'l',
  [Glagolitic.Myslite]: 'm',
  [Glagolitic.Latinate_Myslite]: 'm',
  [Glagolitic.Nashi]: 'n',
  [Glagolitic.Onu]: 'o',
  [Glagolitic.Yeru]: 'ȯ',
  [Glagolitic.Pokoji]: 'p',
  [Glagolitic.Ritsi]: 'r',
  [Glagolitic.Slovo]: 's',
  [Glagolitic.Sha]: 'š',
  [Glagolitic.Tvrido]: 't',
  [Glagolitic.Fita]: 'θ',
  [Glagolitic.Uku]: 'u',
  [Glagolitic.Big_Yus]: 'ų',
  [Glagolitic.Izhitsa]: 'ù',
  [Glagolitic.Vedi]: 'v',
  [Glagolitic.Yeri + Glagolitic.Izhe]: 'y',
  [Glagolitic.Yeru + Glagolitic.Izhe]: 'y',
  [Glagolitic.Zemlja]: 'z',
  [Glagolitic.Zhivete]: 'ž',
});

function glag2latn4word(word: string): string {
  return softener(yerify(deglagolify(word)));
}

export function glag2latn(text: string): string {
  return punctuate(replaceByWord(text, glag2latn4word));
}
