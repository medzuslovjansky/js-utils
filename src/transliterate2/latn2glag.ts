// noinspection NonAsciiCharacters,JSNonASCIINames

import { createReplacer } from './createReplacer';
import { Glagolitic } from '../constants';
import { replaceByWord } from './replaceByWord';
import { syllabicR } from './syllabicR';

export interface Latn2GlagOptions {
  /** @default false */
  latinateMyslite?: boolean;
  /** @default true */
  shta?: boolean;
}

function optionsToInt({ latinateMyslite, shta }: Required<Latn2GlagOptions>) {
  const b0 = latinateMyslite ? 1 : 0;
  const b1 = shta ? 2 : 0;

  return b0 | b1;
}

type Glagolitizer = (word: string) => string;

const MEMO: Record<number, Glagolitizer> = {};

const desoftener = createReplacer({
  ď: 'dь',
  ľ: 'lь',
  lj: 'lь',
  ń: 'nь',
  nj: 'nь',
  ŕ: 'rь',
  ś: 'sь',
  ť: 'tь',
  ź: 'zь',
  Ŕ: 'ьr',
});

const punctuate = createReplacer({
  '?': ';',
  '.': '։',
  ',': '·',
  '(': '⁖',
  ')': '჻',
  '"': '⸪',
  '«': '⸪',
  '»': '⸪',
  '!': '··',
  ':': '⸬',
  ';': '⁘',
  '...': '···',
  '…': '···',
});

const izhefy = createReplacer({
  [Glagolitic.Yeri + Glagolitic.Izhe]:
    Glagolitic.Yeri + Glagolitic.Initial_Izhe,
  [Glagolitic.Izhe + Glagolitic.Izhe]:
    Glagolitic.Izhe + Glagolitic.Initial_Izhe,
});

function buildGlagolitizer({
  latinateMyslite = false,
  shta = false,
}: Latn2GlagOptions) {
  const id = optionsToInt({ latinateMyslite, shta });
  if (!MEMO[id]) {
    const glagolify = createReplacer({
      a: Glagolitic.Azu, // ⰰ
      å: Glagolitic.Otu, // ⱉ
      b: Glagolitic.Buky, // ⰱ
      c: Glagolitic.Tsi, // ⱌ
      ć: shta ? Glagolitic.Shta : Glagolitic.Chrivi + Glagolitic.Yeri, // ⱋ | ⱍⱐ
      č: Glagolitic.Chrivi, // ⱍ
      d: Glagolitic.Dobro, // ⰴ
      đ: Glagolitic.Djervi, // ⰼ
      ѕ: Glagolitic.Dzelo, // ⰷ
      dž: Glagolitic.Dobro + Glagolitic.Zhivete, // ⰴⰶ
      e: Glagolitic.Yestu, // ⰵ
      ě: Glagolitic.Yati, // ⱑ
      ę: Glagolitic.Small_Yus, // ⱔ
      ė: Glagolitic.Yeri, // ⱐ
      ь: Glagolitic.Yeri, // ⰴⱐ
      f: Glagolitic.Fritu, // ⱇ
      g: Glagolitic.Glagoli, // ⰳ
      h: Glagolitic.Heru, // ⱈ
      i: Glagolitic.I, // ⰻ
      j: Glagolitic.Izhe, // ⰹ
      '’j': Glagolitic.Initial_Izhe, // ⰺ
      ję: Glagolitic.Iotated_Small_Yus, // ⱗ
      jo: Glagolitic.Yo, // ⱖ
      ju: Glagolitic.Yu, // ⱓ
      jų: Glagolitic.Iotated_Big_Yus, // ⱙ
      k: Glagolitic.Kako, // ⰽ
      l: Glagolitic.Ljudije, // ⰾ
      m: latinateMyslite ? Glagolitic.Latinate_Myslite : Glagolitic.Myslite, // ⰿ
      mj:
        (latinateMyslite ? Glagolitic.Latinate_Myslite : Glagolitic.Myslite) +
        Glagolitic.Initial_Izhe, // ⰿⰺ
      n: Glagolitic.Nashi, // ⱀ
      o: Glagolitic.Onu, // ⱁ
      ȯ: Glagolitic.Yeru, // ⱏ
      p: Glagolitic.Pokoji, // ⱂ
      pj: Glagolitic.Pokoji + Glagolitic.Initial_Izhe, // ⱂⰺ
      r: Glagolitic.Ritsi, // ⱃ
      R: Glagolitic.Yeru + Glagolitic.Ritsi, // ⱏⱃ
      s: Glagolitic.Slovo, // ⱄ
      sj: Glagolitic.Slovo + Glagolitic.Initial_Izhe, // ⱄⰺ
      š: Glagolitic.Sha, // ⱎ
      t: Glagolitic.Tvrido, // ⱅ
      tj: Glagolitic.Tvrido + Glagolitic.Initial_Izhe, // ⱅⰺ
      θ: Glagolitic.Fita, // ⱚ
      u: Glagolitic.Uku, // ⱆ
      ų: Glagolitic.Big_Yus, // ⱘ
      ù: Glagolitic.Izhitsa, // ⱛ
      v: Glagolitic.Vedi, // ⰲ
      vj: Glagolitic.Vedi + Glagolitic.Initial_Izhe, // ⰲⰺ
      y: Glagolitic.Yeru + Glagolitic.Izhe, // ⱏⰹ
      z: Glagolitic.Zemlja, // ⰸ
      ž: Glagolitic.Zhivete, // ⰶ
    });

    MEMO[id] = (word) => izhefy(glagolify(desoftener(syllabicR(word))));
  }

  return MEMO[id]!;
}

const DEFAULT_OPTIONS: Required<Latn2GlagOptions> = {
  latinateMyslite: false,
  shta: true,
};

export function latn2glag(text: string, options = DEFAULT_OPTIONS): string {
  return punctuate(replaceByWord(text, buildGlagolitizer(options)));
}
