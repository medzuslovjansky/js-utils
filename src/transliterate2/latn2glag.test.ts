import { latn2glag } from './latn2glag';
import { glag2latn } from './glag2latn';

describe('Latn → Glag', () => {
  const SANITY = `\
Na vȯzvyšenosti ovca, ktora ne iměla vȯlnų, uviděla konjev. Pŕvy tęgal tęžky voz, vtory nosil veliko brěmę, tretji brzo vozil mųža.
Ovca rěkla konjam: «Boli mně sŕdce, kȯgda viđų, kako člověk vladaje konjami.»
Konji rěkli: «Slušaj, ovco, nam boli sŕdce, kȯgda vidimo ovo: mųž, gospodaŕ, bere tvojų vȯlnų, da by iměl dlja sebe teplo paĺto. A ovca jest bez vȯlny.»
Uslyšavši to, ovca izběgla v råvninų. | Oficialny noćny odjezd do Kyjeva. T́ma, i korenje revenja počęli råsteńje, a slugi pověděli krålju o veseĺju. `;

  test('sanity check (1)', () => {
    expect(latn2glag(SANITY)).toMatchSnapshot();
  });

  test('sanity check (2)', () => {
    expect(
      latn2glag(SANITY, {
        latinateMyslite: true,
        shta: false,
      }),
    ).toMatchSnapshot();
  });

  test('double check', () => {
    const glag1 = latn2glag(SANITY);
    const glag2 = latn2glag(glag2latn(glag1));
    expect(glag2).toBe(glag1);
  });
});
