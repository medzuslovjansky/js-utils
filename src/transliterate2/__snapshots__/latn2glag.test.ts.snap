// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Latn → Glag sanity check (1) 1`] = `
"Ⱀⰰ ⰲⱏⰸⰲⱏⰹⱎⰵⱀⱁⱄⱅⰻ ⱁⰲⱌⰰ· ⰽⱅⱁⱃⰰ ⱀⰵ ⰻⰿⱑⰾⰰ ⰲⱏⰾⱀⱘ· ⱆⰲⰻⰴⱑⰾⰰ ⰽⱁⱀⱐⰵⰲ։ Ⱂⱐⱃⰲⱏⰹ ⱅⱔⰳⰰⰾ ⱅⱔⰶⰽⱏⰹ ⰲⱁⰸ· ⰲⱅⱁⱃⱏⰹ ⱀⱁⱄⰻⰾ ⰲⰵⰾⰻⰽⱁ ⰱⱃⱑⰿⱔ· ⱅⱃⰵⱅⰺⰻ ⰱⱏⱃⰸⱁ ⰲⱁⰸⰻⰾ ⰿⱘⰶⰰ։
Ⱁⰲⱌⰰ ⱃⱑⰽⰾⰰ ⰽⱁⱀⱐⰰⰿ⸬ ⸪Ⰱⱁⰾⰻ ⰿⱀⱑ ⱄⱐⱃⰴⱌⰵ· ⰽⱏⰳⰴⰰ ⰲⰻⰼⱘ· ⰽⰰⰽⱁ ⱍⰾⱁⰲⱑⰽ ⰲⰾⰰⰴⰰⰹⰵ ⰽⱁⱀⱐⰰⰿⰻ։⸪
Ⰽⱁⱀⱐⰻ ⱃⱑⰽⰾⰻ⸬ ⸪Ⱄⰾⱆⱎⰰⰹ· ⱁⰲⱌⱁ· ⱀⰰⰿ ⰱⱁⰾⰻ ⱄⱐⱃⰴⱌⰵ· ⰽⱏⰳⰴⰰ ⰲⰻⰴⰻⰿⱁ ⱁⰲⱁ⸬ ⰿⱘⰶ· ⰳⱁⱄⱂⱁⰴⰰⱃⱐ· ⰱⰵⱃⰵ ⱅⰲⱁⱙ ⰲⱏⰾⱀⱘ· ⰴⰰ ⰱⱏⰹ ⰻⰿⱑⰾ ⰴⰾⱐⰰ ⱄⰵⰱⰵ ⱅⰵⱂⰾⱁ ⱂⰰⰾⱐⱅⱁ։ Ⰰ ⱁⰲⱌⰰ ⰹⰵⱄⱅ ⰱⰵⰸ ⰲⱏⰾⱀⱏⰹ։⸪
Ⱆⱄⰾⱏⰹⱎⰰⰲⱎⰻ ⱅⱁ· ⱁⰲⱌⰰ ⰻⰸⰱⱑⰳⰾⰰ ⰲ ⱃⱉⰲⱀⰻⱀⱘ։ | Ⱁⱇⰻⱌⰻⰰⰾⱀⱏⰹ ⱀⱁⱋⱀⱏⰹ ⱁⰴⰹⰵⰸⰴ ⰴⱁ Ⰽⱏⰹⰺⰵⰲⰰ։ Ⱅⱐⰿⰰ· ⰻ ⰽⱁⱃⰵⱀⱐⰵ ⱃⰵⰲⰵⱀⱐⰰ ⱂⱁⱍⱔⰾⰻ ⱃⱉⱄⱅⰵⱀⱐⰺⰵ· ⰰ ⱄⰾⱆⰳⰻ ⱂⱁⰲⱑⰴⱑⰾⰻ ⰽⱃⱉⰾⱐⱆ ⱁ ⰲⰵⱄⰵⰾⱐⱓ։ "
`;

exports[`Latn → Glag sanity check (2) 1`] = `
"Ⱀⰰ ⰲⱏⰸⰲⱏⰹⱎⰵⱀⱁⱄⱅⰻ ⱁⰲⱌⰰ· ⰽⱅⱁⱃⰰ ⱀⰵ ⰻⱞⱑⰾⰰ ⰲⱏⰾⱀⱘ· ⱆⰲⰻⰴⱑⰾⰰ ⰽⱁⱀⱐⰵⰲ։ Ⱂⱐⱃⰲⱏⰹ ⱅⱔⰳⰰⰾ ⱅⱔⰶⰽⱏⰹ ⰲⱁⰸ· ⰲⱅⱁⱃⱏⰹ ⱀⱁⱄⰻⰾ ⰲⰵⰾⰻⰽⱁ ⰱⱃⱑⱞⱔ· ⱅⱃⰵⱅⰺⰻ ⰱⱏⱃⰸⱁ ⰲⱁⰸⰻⰾ ⱞⱘⰶⰰ։
Ⱁⰲⱌⰰ ⱃⱑⰽⰾⰰ ⰽⱁⱀⱐⰰⱞ⸬ ⸪Ⰱⱁⰾⰻ ⱞⱀⱑ ⱄⱐⱃⰴⱌⰵ· ⰽⱏⰳⰴⰰ ⰲⰻⰼⱘ· ⰽⰰⰽⱁ ⱍⰾⱁⰲⱑⰽ ⰲⰾⰰⰴⰰⰹⰵ ⰽⱁⱀⱐⰰⱞⰻ։⸪
Ⰽⱁⱀⱐⰻ ⱃⱑⰽⰾⰻ⸬ ⸪Ⱄⰾⱆⱎⰰⰹ· ⱁⰲⱌⱁ· ⱀⰰⱞ ⰱⱁⰾⰻ ⱄⱐⱃⰴⱌⰵ· ⰽⱏⰳⰴⰰ ⰲⰻⰴⰻⱞⱁ ⱁⰲⱁ⸬ ⱞⱘⰶ· ⰳⱁⱄⱂⱁⰴⰰⱃⱐ· ⰱⰵⱃⰵ ⱅⰲⱁⱙ ⰲⱏⰾⱀⱘ· ⰴⰰ ⰱⱏⰹ ⰻⱞⱑⰾ ⰴⰾⱐⰰ ⱄⰵⰱⰵ ⱅⰵⱂⰾⱁ ⱂⰰⰾⱐⱅⱁ։ Ⰰ ⱁⰲⱌⰰ ⰹⰵⱄⱅ ⰱⰵⰸ ⰲⱏⰾⱀⱏⰹ։⸪
Ⱆⱄⰾⱏⰹⱎⰰⰲⱎⰻ ⱅⱁ· ⱁⰲⱌⰰ ⰻⰸⰱⱑⰳⰾⰰ ⰲ ⱃⱉⰲⱀⰻⱀⱘ։ | Ⱁⱇⰻⱌⰻⰰⰾⱀⱏⰹ ⱀⱁⱍⱐⱀⱏⰹ ⱁⰴⰹⰵⰸⰴ ⰴⱁ Ⰽⱏⰹⰺⰵⰲⰰ։ Ⱅⱐⱞⰰ· ⰻ ⰽⱁⱃⰵⱀⱐⰵ ⱃⰵⰲⰵⱀⱐⰰ ⱂⱁⱍⱔⰾⰻ ⱃⱉⱄⱅⰵⱀⱐⰺⰵ· ⰰ ⱄⰾⱆⰳⰻ ⱂⱁⰲⱑⰴⱑⰾⰻ ⰽⱃⱉⰾⱐⱆ ⱁ ⰲⰵⱄⰵⰾⱐⱓ։ "
`;
